File: ts/test/config.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import { writeFileSync, mkdtempSync } from "node:fs";
import { join } from "node:path";
import { tmpdir } from "node:os";

import { loadConfig, defaultSource, source, portFromAddr } from "../src/config.js";

const TOML = `
# dopdb config
[http]
addr           = ":9000"
jwt_secret_env = "CFG_SECRET"
cors_origins   = ["https://a.example.com", "https://b.example.com"]

[[kvrocks]]
name         = "default"
uri_env      = "CFG_KVROCKS"
uri          = "redis://localhost:6666"
password_env = "CFG_KVROCKS_PW"
namespace    = "appdb"

[[kvrocks]]
name      = "analytics"
uri       = "redis://localhost:6667"  # inline comment
namespace = "analytics"
`;

function writeTmp(body: string, name = "config.toml"): string {
  const dir = mkdtempSync(join(tmpdir(), "dopdb-cfg-"));
  const p = join(dir, name);
  writeFileSync(p, body);
  return p;
}

test("TOML: parses http + [[kvrocks]] tables and arrays", () => {
  delete process.env.CFG_KVROCKS;
  delete process.env.CFG_KVROCKS_PW;
  process.env.CFG_SECRET = "s";
  const cfg = loadConfig(writeTmp(TOML));
  assert.equal(cfg.http.addr, ":9000");
  assert.deepEqual(cfg.http.corsOrigins, ["https://a.example.com", "https://b.example.com"]);
  assert.equal(cfg.kvrocks.length, 2);
  assert.equal(defaultSource(cfg).namespace, "appdb");
  assert.equal(source(cfg, "analytics")?.uri, "redis://localhost:6667"); // comment stripped
});

test("env resolution: jwt secret, uri and password come from env, not the file", () => {
  process.env.CFG_SECRET = "from-env";
  process.env.CFG_KVROCKS = "redis://prod-host:6666";
  process.env.CFG_KVROCKS_PW = "ns-token";
  const cfg = loadConfig(writeTmp(TOML));
  assert.equal(cfg.http.jwtSecret, "from-env");
  assert.equal(defaultSource(cfg).uri, "redis://prod-host:6666"); // env wins over literal
  assert.equal(defaultSource(cfg).password, "ns-token");
  delete process.env.CFG_KVROCKS;
  delete process.env.CFG_KVROCKS_PW;
});

test("validate: a default source is required", () => {
  process.env.CFG_SECRET = "s";
  const body = `
[http]
jwt_secret_env = "CFG_SECRET"
[[kvrocks]]
name      = "analytics"
uri       = "redis://x:6666"
namespace = "a"
`;
  assert.throws(() => loadConfig(writeTmp(body)), /default/);
});

// A mongodb:// URI is now a configuration error rather than a misconfiguration
// that only surfaces at dial time.
test("validate: the uri must be redis:// or rediss://", () => {
  process.env.CFG_SECRET = "s";
  const body = `
[http]
jwt_secret_env = "CFG_SECRET"
[[kvrocks]]
name      = "default"
uri       = "mongodb://localhost:27017"
namespace = "appdb"
`;
  assert.throws(() => loadConfig(writeTmp(body)), /redis:\/\//);
});

test("validate: a namespace is required", () => {
  process.env.CFG_SECRET = "s";
  const body = `
[http]
jwt_secret_env = "CFG_SECRET"
[[kvrocks]]
name = "default"
uri  = "redis://localhost:6666"
`;
  assert.throws(() => loadConfig(writeTmp(body)), /namespace/);
});

test("JSON config files load with the same shape", () => {
  const p = writeTmp(JSON.stringify({
    http: { addr: ":1", jwt_secret_env: "CFG_SECRET" },
    kvrocks: [{ name: "default", uri: "redis://localhost:6666", namespace: "j" }],
  }), "config.json");
  process.env.CFG_SECRET = "s";
  const cfg = loadConfig(p);
  assert.equal(defaultSource(cfg).namespace, "j");
});

test("portFromAddr parses the Go-style addr forms", () => {
  assert.equal(portFromAddr(":8080"), 8080);
  assert.equal(portFromAddr("0.0.0.0:9000"), 9000);
  assert.equal(portFromAddr("8080"), 8080);
  assert.equal(portFromAddr("nope"), undefined);
});

File: config/config_test.go
package config

import (
	"os"
	"path/filepath"
	"testing"
)

const sample = `
# dopdb config
[http]
addr           = ":9000"
jwt_secret_env = "TEST_JWT_SECRET"
cors_origins   = ["https://a.example.com", "https://b.example.com"]

[[kvrocks]]
name         = "default"
uri_env      = "TEST_KVROCKS_URI"
uri          = "redis://localhost:6666"   # dev fallback
password_env = "TEST_KVROCKS_PASSWORD"
namespace    = "appdb"

[[kvrocks]]
name      = "analytics"
uri       = "redis://localhost:6666"
namespace = "analytics"
`

func writeTmp(t *testing.T, body string) string {
	t.Helper()
	p := filepath.Join(t.TempDir(), "config.toml")
	if err := os.WriteFile(p, []byte(body), 0o600); err != nil {
		t.Fatal(err)
	}
	return p
}

func TestLoadAndEnvOverride(t *testing.T) {
	t.Setenv("TEST_JWT_SECRET", "s3cr3t")
	t.Setenv("TEST_KVROCKS_URI", "redis://prod:6666")
	t.Setenv("TEST_KVROCKS_PASSWORD", "ns-token")

	cfg, err := Load(writeTmp(t, sample))
	if err != nil {
		t.Fatalf("Load: %v", err)
	}
	if cfg.HTTP.Addr != ":9000" {
		t.Errorf("addr=%q", cfg.HTTP.Addr)
	}
	if cfg.HTTP.JWTSecret != "s3cr3t" {
		t.Errorf("jwt secret not resolved from env: %q", cfg.HTTP.JWTSecret)
	}
	if len(cfg.HTTP.CORSOrigins) != 2 || cfg.HTTP.CORSOrigins[0] != "https://a.example.com" {
		t.Errorf("cors=%v", cfg.HTTP.CORSOrigins)
	}
	def := cfg.Default()
	if def.URI != "redis://prod:6666" {
		t.Errorf("default uri not overridden by env: %q", def.URI)
	}
	if def.Password != "ns-token" {
		t.Errorf("password not resolved from env: %q", def.Password)
	}
	if def.Namespace != "appdb" {
		t.Errorf("default namespace=%q", def.Namespace)
	}
	if _, ok := cfg.Source("analytics"); !ok {
		t.Error("analytics source missing")
	}
}

func TestEnvFallbackToLiteral(t *testing.T) {
	// env not set -> literal uri is used
	t.Setenv("TEST_JWT_SECRET", "x")
	cfg, err := Load(writeTmp(t, sample))
	if err != nil {
		t.Fatalf("Load: %v", err)
	}
	if cfg.Default().URI != "redis://localhost:6666" {
		t.Errorf("expected literal uri fallback, got %q", cfg.Default().URI)
	}
}

func TestValidateRequiresDefault(t *testing.T) {
	t.Setenv("TEST_JWT_SECRET", "x")
	body := `
[http]
jwt_secret_env = "TEST_JWT_SECRET"
[[kvrocks]]
name      = "analytics"
uri       = "redis://localhost:6666"
namespace = "analytics"
`
	if _, err := Load(writeTmp(t, body)); err == nil {
		t.Fatal("expected error: missing default datasource")
	}
}

func TestValidateRequiresSecret(t *testing.T) {
	body := `
[http]
jwt_secret_env = "DEFINITELY_UNSET_VAR_XYZ"
[[kvrocks]]
name      = "default"
uri       = "redis://localhost:6666"
namespace = "appdb"
`
	if _, err := Load(writeTmp(t, body)); err == nil {
		t.Fatal("expected error: empty jwt secret")
	}
}

// A mongodb:// URI is now a configuration error rather than a silent
// misconfiguration that only fails at dial time.
func TestValidateRejectsNonRedisURI(t *testing.T) {
	t.Setenv("TEST_JWT_SECRET", "x")
	body := `
[http]
jwt_secret_env = "TEST_JWT_SECRET"
[[kvrocks]]
name      = "default"
uri       = "mongodb://localhost:27017"
namespace = "appdb"
`
	if _, err := Load(writeTmp(t, body)); err == nil {
		t.Fatal("expected error: uri must be redis:// or rediss://")
	}
}

func TestValidateRequiresNamespace(t *testing.T) {
	t.Setenv("TEST_JWT_SECRET", "x")
	body := `
[http]
jwt_secret_env = "TEST_JWT_SECRET"
[[kvrocks]]
name = "default"
uri  = "redis://localhost:6666"
`
	if _, err := Load(writeTmp(t, body)); err == nil {
		t.Fatal("expected error: missing namespace")
	}
}

func TestWarnings(t *testing.T) {
	cfg := &Config{
		HTTP: HTTPConfig{JWTSecret: "x"},
		Kvrocks: []KvrocksSource{
			{Name: "default", URI: "redis://user:pw@h:6666", Namespace: "d"}, // creds in literal
		},
	}
	w := cfg.Warnings()
	if len(w) != 1 {
		t.Fatalf("expected 1 warning, got %d: %v", len(w), w)
	}
}

func TestWarnsOnLiteralPassword(t *testing.T) {
	cfg := &Config{
		HTTP: HTTPConfig{JWTSecret: "x"},
		Kvrocks: []KvrocksSource{
			{Name: "default", URI: "redis://h:6666", Password: "hunter2", Namespace: "d"},
		},
	}
	if w := cfg.Warnings(); len(w) != 1 {
		t.Fatalf("expected 1 warning for a literal password, got %d: %v", len(w), w)
	}
}

func TestStripCommentRespectsQuotes(t *testing.T) {
	// a '#' inside a quoted value must survive
	body := `
[http]
jwt_secret_env = "TEST_JWT_SECRET"
addr = "host#notacomment:80"   # real comment
[[kvrocks]]
name      = "default"
uri       = "redis://localhost:6666"
namespace = "appdb"
`
	t.Setenv("TEST_JWT_SECRET", "x")
	cfg, err := Load(writeTmp(t, body))
	if err != nil {
		t.Fatalf("Load: %v", err)
	}
	if cfg.HTTP.Addr != "host#notacomment:80" {
		t.Errorf("addr=%q (comment stripping ate a quoted #)", cfg.HTTP.Addr)
	}
}
